### PR TITLE
Add support for asynchronous signing in KeyBox

### PR DESCRIPTION
**Pull request by michalseweryn**
Modify the `KeyBox` trait so that the `sign` method is `async`. This is because blocking signing in deprecated in substrate.

### CODE REVIEW

File: src/member.rs [hunk 1]
```diff
@@ -725,7 +725,7 @@ where
         loop {
             futures::select! {
                 notification = rx_consensus.next() => match notification {
-                        Some(notification) => self.on_consensus_notification(notification),
+                        Some(notification) => self.on_consensus_notification(notification).await,
```
Comment (michalseweryn) [inline src/member.rs:728]: I am not sure if we want to `await` here, we could alternatively add a vector of futures outputting signed data and poll that vector.

File: src/signed.rs [hunk 2]
```diff
@@ -17,10 +17,11 @@ impl<T: Debug + Clone + Encode + Decode + Send + 'static> Signature for T {}
...
 /// and `verify(msg, s, j)` is to verify whether the signature s under the message msg is
 /// correct with respect to the public key of the jth node.
+#[async_trait]
 pub trait KeyBox: Index + Clone + Send + Sync + 'static {
     type Signature: Signature;
     /// Signs a message `msg`.
-    fn sign(&self, msg: &[u8]) -> Self::Signature;
+    async fn sign(&self, msg: &[u8]) -> Self::Signature;
     /// Verifies whether a node with `index` correctly signed the message `msg`.
     fn verify(&self, msg: &[u8], sgn: &Self::Signature, index: NodeIndex) -> bool;
```
Comment (timorl) [inline src/signed.rs:26]: If `sign` is already `async` shouldn't `verify` also be? The access the same resources in very similar ways, so it's quite a bit strange that one is and one isn't.

File: src/signed.rs [hunk 3]
```diff
@@ -90,7 +91,7 @@ pub struct UncheckedSigned<T: Signable, S> {
 /// to upgrade this `struct` to [`Signed<'a, T, KB>`] which ensures that the signature matches the
 /// signed object, and the method [`UncheckedSigned::check_multi`] can be used to upgrade to
 /// [`Multisigned<'a, T, MK>`].
-impl<T: Signable, S: Signature> UncheckedSigned<T, S> {
```
Comment (timorl) [inline src/signed.rs:93]: Why did you remove that bound? Is some linter complaining that it technically isn't used?

File: src/signed.rs [hunk 4]
```diff
@@ -192,9 +193,9 @@ impl<'a, T: Signable + Clone + Index, KB: KeyBox> Clone for Signed<'a, T, KB> {
 
 impl<'a, T: Signable + Index, KB: KeyBox> Signed<'a, T, KB> {
     /// Create a signed object from a signable. The index of `signable` must match the index of the `key_box`.
-    pub fn sign(signable: T, key_box: &'a KB) -> Self {
+    pub async fn sign(signable: T, key_box: &'a KB) -> Signed<'a, T, KB> {
```
Comment (timorl) [inline src/signed.rs:196]: Huh, `Self` doesn't work with `async`?

**Review by timorl — CHANGES_REQUESTED** [review]
Some nits and questions and a suggestion about also making `verify` async.

File: src/signed.rs (hunk 2 above)
Comment (michalseweryn) [reply src/signed.rs:26]: In substrate, public keys are stored in-memory while the private keys can be stored in the filesystem (https://crates.parity.io/sc_keystore/struct.LocalKeystore.html), this is why only signing is `async`.

File: src/signed.rs (hunk 3 above)
Comment (michalseweryn) [reply src/signed.rs]: I don't remember why did the compiler complain. In general there is a question whether the  bounds should be in the struct definition or in the `impl` blocks. It seems like the convention in Rust is to put them in the definition, so I restored the bound and propagated it to all `impl` s

File: src/signed.rs (hunk 4 above)
Comment (michalseweryn) [reply src/signed.rs:196]: The compiler complains when `Self` has lifetimes
